core/src/main/java/hudson/model/Queue.java

```diff
@@ -1487,6 +1519,8 @@ public void maintain() {
                             // determine if they are blocked by the lucky winner
                             updateSnapshot();
                         }
+                    } else {
+                        p.setCauseOfBlockage(causeOfBlockage);
```

Comment:
if ``allowNewBuildableTask` returns false, a non-null value may be overridden by a null one here. Is it "as designed"?

Reply:
Good catch. Moved `allowNewBuildableTask` to  `getCauseOfBlockage` as well.